File: DAA practical codes/DAA python/mergbus.cpp
/*
---------------------------------------------------------
      MERGE SORT - BUS TICKET PRICE SORTING
---------------------------------------------------------

SAMPLE INPUT:
Enter number of ticket prices: 5
Enter ticket prices: 120 45 60 30 250

SAMPLE OUTPUT:
Sorted ticket prices (cheapest first):
30 45 60 120 250
---------------------------------------------------------
*/

#include <iostream>
using namespace std;

void mergeArr(int a[], int l, int m, int r) {
    int n1 = m - l + 1;
    int n2 = r - m;

    int left[100], right[100];

    for (int i = 0; i < n1; i++)
        left[i] = a[l + i];
    for (int j = 0; j < n2; j++)
        right[j] = a[m + 1 + j];

    int i = 0, j = 0, k = l;

    while (i < n1 && j < n2) {
        if (left[i] <= right[j]) {
            a[k] = left[i];
            i++;
        } else {
            a[k] = right[j];
            j++;
        }
        k++;
    }

    while (i < n1) {
        a[k] = left[i];
        i++;
        k++;
    }

    while (j < n2) {
        a[k] = right[j];
        j++;
        k++;
    }
}

void mergeSort(int a[], int l, int r) {
    if (l < r) {
        int m = (l + r) / 2;
        mergeSort(a, l, m);
        mergeSort(a, m + 1, r);
        mergeArr(a, l, m, r);
    }
}

int main() {
    int n;
    cout << "Enter number of ticket prices: ";
    cin >> n;

    int price[100];

    cout << "Enter ticket prices: ";
    for (int i = 0; i < n; i++)
        cin >> price[i];

    mergeSort(price, 0, n - 1);

    cout << "\nSorted ticket prices (cheapest first):\n";
    for (int i = 0; i < n; i++)
        cout << price[i] << " ";

    return 0;
}

File: DAA practical codes/DAA python/mergbus.py
"""
---------------------------------------------------------
      MERGE SORT - BUS TICKET PRICE SORTING
---------------------------------------------------------

SAMPLE INPUT:
Enter number of ticket prices: 5
Enter ticket prices: 120 45 60 30 250

SAMPLE OUTPUT:
Sorted ticket prices (cheapest first):
30 45 60 120 250
---------------------------------------------------------
"""

def merge(arr, l, m, r):
    left = arr[l:m+1]
    right = arr[m+1:r+1]

    i = j = 0
    k = l

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            arr[k] = left[i]
            i += 1
        else:
            arr[k] = right[j]
            j += 1
        k += 1

    while i < len(left):
        arr[k] = left[i]
        i += 1
        k += 1

    while j < len(right):
        arr[k] = right[j]
        j += 1
        k += 1

def mergeSort(arr, l, r):
    if l < r:
        m = (l + r) // 2
        mergeSort(arr, l, m)
        mergeSort(arr, m + 1, r)
        merge(arr, l, m, r)

print("Enter number of ticket prices:", end=" ")
n = int(input())

print("Enter ticket prices:", end=" ")
data = list(map(int, input().split()))

mergeSort(data, 0, n - 1)

print("\nSorted ticket prices (cheapest first):")
for x in data:
    print(x, end=" ")
